src/commands/test/lib/test-cloud-uploader.ts

```diff
@@ -123,52 +124,44 @@ export class TestCloudUploader {
     });
   }
 
-  private async uploadHashOrNewFile(testRunId: string, file: TestRunFile): Promise<void> {
-    if (await this.tryUploadFileHash(testRunId, file)) {
-      debug(`File ${file.sourcePath}: hash upload`);
-    }
-    else {
-      await this.uploadFile(testRunId, file);
-      debug(`File ${file.sourcePath}: direct upload`);
-    }
+  private async uploadFilesUsingBatch(testRunId: string, files: TestRunFile[]): Promise<void> {
+    let checkHashResult = await this.uploadHashesBatch(testRunId, files.map(f => { return { file: f }; }));
+
+    let limit = pLimit(paralleRequests);
+    let uploadNewFilesTask = checkHashResult
+      .filter(r => r.response.uploadStatus.statusCode === 412)
+      .map(r => limit(() => this.uploadFile(testRunId, r.file)));
+
+    await Promise.all(uploadNewFilesTask);
   }
 
-  private async tryUploadFileHash(testRunId: string, file: TestRunFile, byteRange: string = null): Promise<boolean> {
-    let response = await new Promise<http.IncomingMessage>((resolve, reject) => {
-      this._client.test.uploadHash(
+  private async uploadHashesBatch(testRunId: string, files: { file: TestRunFile, byteRange?: string }[]): Promise<{ file: TestRunFile, response: models.TestCloudFileHashResponse }[]> { 
+    let hashResponses = await new Promise<models.TestCloudFileHashResponse[]>((resolve, reject) => {
```

Comment:
We have new new `clientRequest` wrapper that gives you access to the actual HTTP response as well as the result, can you use that now instead of manual promise wrangling?